core/shared/utils/bh_atomic.h

```diff
@@ -58,7 +71,37 @@ typedef uint16 bh_atomic_16_t;
 #endif
 
 #if defined(CLANG_GCC_HAS_ATOMIC_BUILTIN)
+#ifndef BH_ATOMIC_32_IS_ATOMIC
 #define BH_ATOMIC_32_IS_ATOMIC 1
+#endif
+#ifndef BH_ATOMIC_16_IS_ATOMIC
+#define BH_ATOMIC_16_IS_ATOMIC 1
+#endif
+#else
+#ifndef BH_ATOMIC_32_IS_ATOMIC
+#define BH_ATOMIC_32_IS_ATOMIC 0
+#endif
+#ifndef BH_ATOMIC_16_IS_ATOMIC
+#define BH_ATOMIC_16_IS_ATOMIC 0
+#endif
+#endif
+
+/* Force disable atomic 16-bit operations on bare-metal RISC-V
+ * because the 16-bit atomic operations is emulated by 32-bit
+ * atomic operations, which has linkage problem on current toolchain:
+ * in function `shared_memory_inc_reference':
+ * wasm_shared_memory.c:85:(.text.shared_memory_inc_reference+0x10): undefined
+ * reference to `__atomic_fetch_add_2'
+ */
+#ifndef WASM_UINT16_IS_ATOMIC
+#if !defined(__linux__) && defined(__riscv)
```

Comment:
what do you mean by `!defined(__linux__)` here?
do you mean `defined(__NuttX__)`?

also, why do you override it here, rather than having it in platform-specific headers?

Reply:
I guess all non-linux/baremetal platforms has the same issue, not only nuttx, since the usage of the baremetal toolchain have nothing difference with other RTOS platform.

And you can also define the behavior in platfrom-specific setting, explicit settings always have the highest priority.

Reply:
how about freebsd?

Reply:
I think the bsd family should be OK as linux.